services/retention/service.go

```diff
@@ -80,83 +103,124 @@ func (s *Service) run() {
-						if err := s.MetaClient.DeleteShardGroup(d.Name, r.Name, g.ID); err != nil {
-							log.Info("Failed to delete shard group",
-								logger.Database(d.Name),
-								logger.ShardGroup(g.ID),
-								logger.RetentionPolicy(r.Name),
-								zap.Error(err))
-							retryNeeded = true
-							continue
-						}
-
-						log.Info("Deleted shard group",
-							logger.Database(d.Name),
-							logger.ShardGroup(g.ID),
-							logger.RetentionPolicy(r.Name))
-
-						// Store all the shard IDs that may possibly need to be removed locally.
-						for _, sh := range g.Shards {
-							deletedShardIDs[sh.ID] = deletionInfo{db: d.Name, rp: r.Name}
-						}
-					}
+
+			// Determine all shards that have expired and need to be deleted.
+			for _, g := range r.ExpiredShardGroups(time.Now().UTC()) {
+				if err := s.MetaClient.DeleteShardGroup(d.Name, r.Name, g.ID); err != nil {
+					log.Info("Failed to delete shard group",
+						logger.Database(d.Name),
+						logger.ShardGroup(g.ID),
+						logger.RetentionPolicy(r.Name),
+						zap.Error(err))
+					retryNeeded = true
+					continue
+				}
+
+				log.Info("Deleted shard group",
+					logger.Database(d.Name),
+					logger.ShardGroup(g.ID),
+					logger.RetentionPolicy(r.Name))
+
+				// Store all the shard IDs that may possibly need to be removed locally.
+				for _, sh := range g.Shards {
+					deletedShardIDs[sh.ID] = deletionInfo{db: d.Name, rp: r.Name}
 				}
 			}
+		}
+	}
 
-			// Remove shards if we store them locally
-			for _, id := range s.TSDBStore.ShardIDs() {
-				if info, ok := deletedShardIDs[id]; ok {
-					if err := s.TSDBStore.DeleteShard(id); err != nil {
-						log.Info("Failed to delete shard",
-							logger.Database(info.db),
-							logger.Shard(id),
-							logger.RetentionPolicy(info.rp),
-							zap.Error(err))
-						retryNeeded = true
-						continue
-					}
-					log.Info("Deleted shard",
+	// Remove shards if we store them locally
+	for _, id := range s.TSDBStore.ShardIDs() {
+		if info, ok := deletedShardIDs[id]; ok {
+			delete(deletedShardIDs, id)
+			if err := s.TSDBStore.DeleteShard(id); err != nil {
+				log.Error("Failed to delete shard",
+					logger.Database(info.db),
+					logger.Shard(id),
+					logger.RetentionPolicy(info.rp),
+					zap.Error(err))
+				if errors.Is(err, tsdb.ErrShardNotFound) {
+					log.Warn("Shard does not exist in store, continuing retention removal",
 						logger.Database(info.db),
 						logger.Shard(id),
 						logger.RetentionPolicy(info.rp))
+				} else {
+					retryNeeded = true
+					continue
 				}
 			}
-
-			if err := s.MetaClient.PruneShardGroups(); err != nil {
-				log.Info("Problem pruning shard groups", zap.Error(err))
+			log.Info("Deleted shard",
+				logger.Database(info.db),
+				logger.Shard(id),
+				logger.RetentionPolicy(info.rp))
+			if err := dropShardMetaRef(id, info); err != nil {
+				// removeShardMetaReference already logged the error.
 				retryNeeded = true
+				continue
```

Comment:
Does this `continue` do anything?

Reply:
No, it is for symmetry with the code that was already here. If someone comes and adds some code after this `if`, though, the `continue` is important.

I'd probably remove the use of `continue` completely if I was making more drastic changes.